src/examples/algorithms/AnomalyDetection.cpp

```diff
@@ -0,0 +1,221 @@
+   detection using provided stopwatch Timer. 
+
+ Usage:
+   run $RELEASE/bin/anomaly_example 
+ */
+
+#include <iostream>
+#include <fstream> // for file output
+#include <vector>
+#include <string>
+#include <cstdlib>
+
+#include "nupic/encoders/ScalarEncoder.hpp"
+#include "nupic/algorithms/SpatialPooler.hpp"
+#include "nupic/algorithms/Cells4.hpp" // TP
+#include <nupic/algorithms/TemporalMemory.hpp> // TM
+#include "nupic/algorithms/Anomaly.hpp"
+
+#include <csv.h> // external CSV parser
+#include "nupic/os/Timer.hpp"
+#include "nupic/utils/VectorHelpers.hpp"
+
+using namespace std; // generic namespaces are included here
+using namespace nupic;
+using namespace nupic::utils;
+using namespace nupic::algorithms::spatial_pooler;
+using namespace nupic::algorithms::Cells4;
+using namespace nupic::algorithms::temporal_memory;
+using namespace nupic::algorithms::anomaly;
+
+const int DEBUG_LEVEL =1; //0=no debug (also disabled timer), ..
+const string tmImpl = "TM"; //"TM","TP" //TODO make param
+
+class AnomalyDetection
+{
+
+  private:
+    ScalarEncoder encoder;
+    SpatialPooler sp;
+    Cells4 tp;
+    TemporalMemory tm;
+    Anomaly anomaly; 
+    std::vector<UInt> lastTPOutput_;
+
+  public:
+    Real compute(Real x) {
+        auto scalar = std::vector<Real32>(encoder.getOutputWidth());
+        encoder.encodeIntoArray(x, scalar.data());
+
+        // For some reason, the ScalarEncoder outputs a Real32 vector,
+        // which is incomaptible with the SpatialPooler's input, hence a new
+        // vector is created with the required type, and the elements
+        // are converted.
+        auto uint_scalar = VectorHelpers::castVectorType<Real32, UInt>(scalar);
+        if (DEBUG_LEVEL > 1) {
+          std::cout << "Scalar encoder: ";
+          VectorHelpers::print_vector(uint_scalar);
+        }
+
+        std::vector<UInt> spOutput(sp.getNumColumns());
+        sp.compute(uint_scalar.data(), true, spOutput.data());
+        sp.stripUnlearnedColumns(spOutput.data());
+        if (DEBUG_LEVEL > 2) {
+          std::cout << "Spatial pooler: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(spOutput), ",");
+          std::cout << std::endl;
+        }
+
+        // The temporal pooler uses Real32, so the vector must be converted again
+        auto tpInput = VectorHelpers::castVectorType<UInt, Real>(spOutput);
+        auto nCells = 0;
+        if (tmImpl == "TP") {
+          nCells = tp.nCells(); //TODO fix API of (SP)/TM/TP to same format, so this can be avoided
+        else { 
+          nCells = tm.numberOfCells();
+        }
+        std::vector<Real> tpOutput(nCells);
+        if (tmImpl == "TP") {
+          tp.compute(tpInput.data(), tpOutput.data(), true, true);
+        else {
+          tm.compute(tpInput.size(), tpInput.data(), true);
+          tpOutput = tm.getActiveCells(); //FIXME do union with getPredictedCells() , like TP.outputMode="both"
```

Comment:
If you want the union of active and predictive cells, it should be easy to build it using `tm.getActiveCells()` and `tm.getPredictiveCells()`.